### PR TITLE
Crude port of lockfile_unix logic using Windows LockFileEx api

### PR DESCRIPTION
**Pull request by davidvoit**
skopeo is currently not able store container data under windows. The first blocking point is that the lock file implementation under
Windows is only a stub.

This commit ports the Linux logic using LockFileEx Windows API.

Please review this code carefully as I don't have any knowledge about the Windows API and this is my first go code ever touched.
The unit test is passing now, but if the filelocking is correct and every edge case is handled I can't guarantee.

### CODE REVIEW

File: pkg/lockfile/lockfile_windows.go [hunk 1]
```diff
@@ -22,48 +69,190 @@ import (
 // - There may or MAY NOT be an actual object on the filesystem created for the specified path.
 // - Even if ro, the lock MAY be exclusive.
 func createLockerForPath(path string, ro bool) (Locker, error) {
-	return &lockfile{locked: false}, nil
+	// Check if we can open the lock.
+	fd, err := openLock(path, ro)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening %q", path)
```
Comment (rhatdan) [inline pkg/lockfile/lockfile_windows.go:75]: I have a feeling this is a stutter, IE The path would show up in the err from the openLock call above.

File: pkg/lockfile/lockfile_windows.go [hunk 2]
```diff
@@ -22,48 +69,190 @@ import (
...
+	}
+	err = syscall.Close(fd)
```
Comment (rhatdan) [inline pkg/lockfile/lockfile_windows.go:77]: Put the if err :=syscll.Close(fd); err {
}

File: pkg/lockfile/lockfile_windows.go [hunk 3]
```diff
@@ -22,48 +69,190 @@ import (
...
+	default:
+		panic(fmt.Sprintf("attempted to acquire a file lock of unrecognized type %d", lType))
+	}
+	l.stateMutex.Lock()
+	defer l.stateMutex.Unlock()
+	if l.counter == 0 {
+		// If we're the first reference on the lock, we need to open the file again.
+		fd, err := openLock(l.file, l.ro)
+		if err != nil {
+			panic(fmt.Sprintf("error opening %q: %v", l.file, err))
```
Comment (rhatdan) [inline pkg/lockfile/lockfile_windows.go:134]: stutter?

File: pkg/lockfile/lockfile_windows.go [hunk 4]
```diff
@@ -22,48 +69,190 @@ import (
...
+		}
+		l.fd = fd
+
+		var ol syscall.Overlapped
+
+		// Optimization: only use the (expensive) fcntl syscall when
+		// the counter is 0.  In this case, we're either the first
+		// reader lock or a writer lock.
+		ret, _, err := procLockFileEx.Call(uintptr(l.fd), uintptr(lType), 0, uintptr(MAXDWORD), uintptr(MAXDWORD), uintptr(unsafe.Pointer(&ol)))
```
Comment (rhatdan) [inline pkg/lockfile/lockfile_windows.go:143]: Need to do something with this err. At lease log it.

File: pkg/lockfile/lockfile_windows.go [hunk 5]
```diff
@@ -22,48 +69,190 @@ import (
...
 	l.locked = true
+	l.recursive = recursive
+	l.counter++
+}
+
+// Lock locks the lockfile as a writer.  Panic if the lock is a read-only one.
+func (l *lockfile) Lock() {
+	if l.ro {
+		panic("can't take write lock on read-only lock file")
+	} else {
```
Comment (rhatdan) [inline pkg/lockfile/lockfile_windows.go:158]: No need for else. just call l.lock()

File: pkg/lockfile/lockfile_windows.go [hunk 6]
```diff
@@ -22,48 +69,190 @@ import (
...
+		l.lock(LOCKFILE_EXCLUSIVE_LOCK, false)
+	}
 }
 
 func (l *lockfile) RecursiveLock() {
-	// We don't support Windows but a recursive writer-lock in one process-space
-	// is really a writer lock, so just panic.
-	panic("not supported")
+	if l.ro {
+		l.RLock()
```
Comment (rhatdan) [inline pkg/lockfile/lockfile_windows.go:165]: just return l.RLock()
And return l.lock()

File: pkg/lockfile/lockfile_windows.go [hunk 7]
```diff
@@ -22,48 +69,190 @@ import (
...
+	n, err := syscall.Write(l.fd, id)
+	if err != nil {
+		return err
+	}
+	if n != len(id) {
+		return syscall.ENOSPC
+	}
 	return nil
 }
 func (l *lockfile) IsReadWrite() bool {
```
Comment (rhatdan) [inline pkg/lockfile/lockfile_windows.go:254]: Move this to lockfile.go.

File: pkg/lockfile/lockfile_windows.go [hunk 8]
```diff
@@ -22,48 +69,190 @@ import (
...
 func (l *lockfile) RLock() {
-	l.mu.Lock()
-	l.locked = true
+	l.lock(0, false)
 }
 
 func (l *lockfile) Unlock() {
-	l.locked = false
-	l.mu.Unlock()
+	l.stateMutex.Lock()
```
Comment (rhatdan) [inline pkg/lockfile/lockfile_windows.go:176]: The cut and paste code from lockfile_unix.go should be moved to lockfile.go and removed from lockfile_windows.go and lockfile_unix.go.

**Comment by davidvoit** [comment]
Hi Daniel,

Thanks for the review, I will try to work on your requested changes, this
week

Regards,
David

Daniel J Walsh ***@***.***> schrieb am Mo., 27. Dez. 2021,
03:40:

**Comment by rhatdan** [comment]
@davidvoit Any update.  We are looking to cut a release, if you want to get this in?

**Comment by davidvoit** [comment]
Hi Dan,

no don't want to rush this. My golang hacking level is copy paste, not more, I tried to understand the type system to get something like a abstract base class working, which can share the implementation between both OSes.

This didn't work out like expected and I need to learn golang more first. Maybe latter I will sent a new pull request, let's see, I am doing this for fun.
So let's close this pull request for now.
@Anybody else: you are free to build on this pull request if your golang foo is better than mine :-)

Anyways thanks for your feedback and time!
